metaflow/plugins/pypi/pip.py

```diff
@@ -89,21 +109,76 @@ def download(self, id_, packages, python, platform):
         # download packages only if they haven't ever been downloaded before
         if os.path.isfile(metadata_file):
             return
+
         metadata = {}
+        custom_index_url, extra_index_urls = self.indices(prefix)
+
+        # build wheels if needed
+        with ThreadPoolExecutor() as executor:
+
+            def _build(key, package):
+                dest = "{prefix}/.pip/built_wheels/{key}".format(prefix=prefix, key=key)
+                cmd = [
+                    "wheel",
+                    "--no-deps",
+                    "--progress-bar=off",
+                    "--wheel-dir=%s" % dest,
+                    "--quiet",
+                    *(["--index-url", custom_index_url] if custom_index_url else []),
+                    *(
+                        chain.from_iterable(
+                            product(["--extra-index-url"], set(extra_index_urls))
+                        )
+                    ),
+                    package["url"],
+                ]
+                self._call(prefix, cmd)
+                return package, dest
+
+            results = list(
+                executor.map(
+                    lambda x: _build(*x),
+                    enumerate(
+                        package for package in packages if package["require_build"]
+                    ),
+                )
+            )
+            os.makedirs("%s/.pip/wheels" % prefix, exist_ok=True)
+
+            for package, path in results:
+                (wheel,) = [
```

Comment:
this will outright fail if the folder doesn't contain exactly one .whl file. Unsure if that is a relevant case to cover at this point though